README.md
# Websocket with private message and rooms


CHANGELOG.md
# Changelog

### [1.1.2](https://www.github.com/SecretHouseGame/websocket/compare/v1.1.1...v1.1.2) (2022-06-21)


### Bug Fixes

* port websocket ([884a52b](https://www.github.com/SecretHouseGame/websocket/commit/884a52b541a84687eb1590450c1ea600fa9bbb31))

### [1.1.1](https://www.github.com/SecretHouseGame/websocket/compare/v1.1.0...v1.1.1) (2022-06-20)


### Bug Fixes

* change port of websocket server ([7e5dbea](https://www.github.com/SecretHouseGame/websocket/commit/7e5dbeafe13a6b28ec3b54bc0c885212de2c6e42))

## [1.1.0](https://www.github.com/SecretHouseGame/websocket/compare/v1.0.0...v1.1.0) (2022-06-19)


### Features

* edit params order and add roomId ([10ad29c](https://www.github.com/SecretHouseGame/websocket/commit/10ad29c4d19fda10fbe97bb3ff112e70c549ee59))

## 1.0.0 (2022-06-19)


### Features

* add ci to auto-deploy with github action ([#11](https://www.github.com/SecretHouseGame/websocket/issues/11)) ([f5a8094](https://www.github.com/SecretHouseGame/websocket/commit/f5a80944c00d1b3f292fef8c3fe7205133363613))
* can send in private message, but other user can't read message ([605d43e](https://www.github.com/SecretHouseGame/websocket/commit/605d43eac42134d6b543555fdda3dc7372723947))
* private message work ([7290cca](https://www.github.com/SecretHouseGame/websocket/commit/7290cca531063ac82c16e0b0a878a92ffa1b266b))
* starting private message ([a2051c7](https://www.github.com/SecretHouseGame/websocket/commit/a2051c765a4f1ff87182fffc86d2f257d8f96122))


### Bug Fixes

* chat global ([8f0e0d7](https://www.github.com/SecretHouseGame/websocket/commit/8f0e0d7503b671edad502528c4d1c72e871175b8))
* cors to local ([e75ba81](https://www.github.com/SecretHouseGame/websocket/commit/e75ba8149bc7d7c55d3f74fbfcb9329bbc84267a))
* private message ([9cc9b8b](https://www.github.com/SecretHouseGame/websocket/commit/9cc9b8be176791d9782e593b8a627340ca9a01f5))
* truc ([a43a157](https://www.github.com/SecretHouseGame/websocket/commit/a43a1577a716d67f16a0461b7df522d021ce290c))
* username when user is disconnected and add socket.client.id ([8bc7220](https://www.github.com/SecretHouseGame/websocket/commit/8bc722046e254c7708489f5941af62170a00391e))
* variable ([f36750e](https://www.github.com/SecretHouseGame/websocket/commit/f36750ec468173ea22b3dec80ad7b1e8226ffcc9))
* variable, PartyId and private message ([31a9a89](https://www.github.com/SecretHouseGame/websocket/commit/31a9a899377090a45e9bec8ac1231b53ba47b157))


server.js
'use strict';
const express = require("express");
const {createServer} = require("http");
const {Server} = require("socket.io");
const app = express();
const Sentry = require('@sentry/node');
const Tracing = require("@sentry/tracing");
const ent = require('ent');
const httpServer = createServer(app);
const io = new Server(httpServer, {
  cors: {
    origins: [
        'http://localhost:4200',
        'https://www.secrethouse.online'
    ]
  }
});

Sentry.init({
  dsn: "https://ac5eefb263564a459ed5fb089aceecd2@o1064146.ingest.sentry.io/6511891",
  integrations: [
    // enable HTTP calls tracing
    new Sentry.Integrations.Http({ tracing: true }),
    // enable Express.js middleware tracing
    new Tracing.Integrations.Express({ app }),
  ],

  // Set tracesSampleRate to 1.0 to capture 100%
  // of transactions for performance monitoring.
  // We recommend adjusting this value in production
  tracesSampleRate: 1.0,
});

// RequestHandler creates a separate execution context using domains, so that every
// transaction/span/breadcrumb is attached to its own Hub instance
app.use(Sentry.Handlers.requestHandler());
// TracingHandler creates a trace for every incoming request
app.use(Sentry.Handlers.tracingHandler());

app.set('view engine', 'ejs');
const users = [];

app.get('/', (req, res) => {
  res.render('index');
})
app.get('/client', function (req, res) {
  res.sendFile(__dirname + "/client.js")
})
app.get("/debug-sentry", function mainHandler(req, res) {
  throw new Error("My first Sentry error!");
});

io.on("connection", (socket) => {
  // Private message
  socket.on("private message", (partyId, roomId, username, avatar, message, fromUserId, toUserId, type) => {
    socket.broadcast.emit(toUserId, {partyId, username, avatar, message, fromUserId, toUserId, type});
  });

  socket.on('disconnect', function () {
    users.forEach(function (user) {
      if (user.client === socket.client.id) {
        socket.broadcast.emit('message_' + user.partyId + '_' + user.roomId, {message: user.username + ' is disconnected '});

        for (let i=0; i < users.length; i++) {
          if (users[i] === user) {
            users.splice(i, 1);
            break;
          }
        }

      }
    })
  });

  // TODO: add RoomList to send event in every room. (Foreach)
  socket.on('event', function (partyId, roomId, username, avatar, message, fromUserId, toUserId) {
    let type = 'event';
    message = ent.encode(message);

    users.forEach(function (user) {
      if (user.fromUserId === fromUserId && roomId !== user.roomId) {
        socket.broadcast.emit('message_' + user.partyId + '_' + user.roomId, {
          partyId: user.partyId,
          roomId: user.roomId,
          username,
          avatar,
          message: user.username + ' vient de quitter la pièce',
          fromUserId,
          toUserId,
          type
        });
      }
    });

    users.push({username, avatar, fromUserId, partyId, roomId});

    socket.broadcast.emit('message_' + partyId + '_' + roomId,{partyId, roomId, username, avatar, message, fromUserId, toUserId, type});
  });

  // Dès qu'on reçoit un message, on récupère le pseudo de son auteur et on le transmet aux autres personnes
  socket.on('tab-general', function (partyId, roomId, username, avatar, message, fromUserId, toUserId) {
    users.push({username, avatar, fromUserId, partyId, roomId});

    console.log(partyId, roomId, username, avatar, message, fromUserId, toUserId);

    socket.broadcast.emit('message_' + partyId + '_' + roomId,{partyId, roomId, username, avatar, message, fromUserId, toUserId, type: 'general-message'});
  });
});

// The error handler must be before any other error middleware and after all controllers
app.use(Sentry.Handlers.errorHandler());

// Optional fallthrough error handler
app.use(function onError(err, req, res, next) {
  // The error id is attached to `res.sentry` to be returned
  // and optionally displayed to the user for support.
  res.statusCode = 500;
  res.end(res.sentry + "\n");
});

httpServer.listen(3000);



client.js
// Connexion à socket.io
const socket = io.connect('http://localhost:3000');
let currentUser = '';
let party = '';
let userId = '';
let arrayDiscussion = [];
localStorage.setItem('type', 'general')

// On demande le pseudo, on l'envoie au serveur et on l'affiche dans le titre
if (localStorage.getItem('username')) {
    currentUser = localStorage.getItem('username');
} else {
    currentUser = prompt('Quel est votre pseudo ?');
    localStorage.setItem('username', currentUser);
}
if (localStorage.getItem('userId')) {
    userId = localStorage.getItem('userId');
} else {
    userId = prompt('Quel est votre id ?');
    localStorage.setItem('userId', userId);
}

roomId = prompt('Dans quel piece voulez-vous aller ?');
localStorage.setItem('room', roomId);
document.querySelector(`div.fade.card-body[data-room='tab-general']`).id = 'zone-chat-' + roomId;

// Fake party ID
if (localStorage.getItem('party')) {
    party = localStorage.getItem('party');
} else {
    party = prompt('Veuillez saisir l\'id de votre party');
    localStorage.setItem('party', party);
}

localStorage.setItem('channel', 'tab-general');

// When logged on websocket, send user info and get Lobby room Id
document.title = currentUser + ' - ' + document.title;
socket.on('connect', () => {
    //NEW PARAMS DONE
    sendMessage(
        'event',
        localStorage.getItem('party'),
        localStorage.getItem('room'),
        currentUser,
        'avatar',
        currentUser + ' vient de rejoindre la pièce ' + userId,
        userId,
        userId
    )

    socket.on(userId, function (data) {
        if (!arrayDiscussion.includes(data.fromUserId)) {
            arrayDiscussion.push(data.fromUserId)

            insertTab(data.username, data.fromUserId);
            insertChat(data.fromUserId);
        }

        if (localStorage.getItem('channel') !== data.fromUserId) {
            addNotification(data.fromUserId)
        }

        // NEW PARAMS DONE
        insertMessage(
            localStorage.getItem('channel'),
            data.partyId,
            data.roomId,
            data.username,
            data.avatar,
            data.message,
            data.fromUserId,
            data.toUserId ?? data.fromUserId,
            data.type
        )
    })

    socket.on("message_" + party + '_' + roomId, function (data) {
        if (localStorage.getItem('channel') !== 'tab-general') {
            addNotification('tab-general')
        }

        // NEW PARAMS DONE
        insertMessage(
            localStorage.getItem('channel'),
            data.partyId,
            data.roomId,
            data.username,
            data.avatar,
            data.message,
            data.fromUserId,
            data.toUserId ?? data.fromUserId,
            data.type
        )
    })
});

// NEW PARAMS DONE
function sendMessage(channel, partyId, roomId, username, avatar, message, fromUserId, toUserId) {
    socket.emit(channel, localStorage.getItem('party'), roomId, username, avatar, message, fromUserId, toUserId); // Transmet le message aux autres
}

function addNotification(tab) {
    const el = document.querySelector(`button.nav-link#` + tab);
    const notification = document.querySelector(`div#notification-` + tab);

    if (notification) {
        let value = notification.textContent;
        value++;
        notification.innerHTML = value;
    } else {
        el.insertAdjacentHTML('beforeend', `
        <div id="notification-${tab}" class="notifications bg-red-500 rounded-lg px-3 object-fill ml-1">1</div>
    `)
    }
}

// Add message in chat page
// NEW PARAMS DONE
function insertMessage(channel, partyId, roomId, username, avatar, message, fromUserId, toUserId, type) {
    let chat = '';

    if (channel === 'tab-general') {
        chat = document.querySelector('#zone-chat-' + roomId);
    } else {
        chat = document.querySelector('#zone-chat-' + channel);
    }

    if (type === "event") {
        chat.insertAdjacentHTML(
            'beforeend',
            '<div class="justify-center text-xs p-3">' + message + '</div>');
    } else {
        chat.insertAdjacentHTML(
            'beforeend',
            '<div><a href="" onclick="getUserInfo(this, `' + username + '`, `' + fromUserId + '`)" data-action="infos-user" data-id=' + fromUserId +
            '   data-type="private_message" data-channel_id="' + fromUserId + '" data-bs-toggle="pill" data-bs-target="#zone-chat-' + fromUserId + '" role="tab" aria-controls="pills-profile" aria-selected="false"  ' +
            '>' + username + '</a> ' + message + '</div>');
    }

    chat.scroll({
        top: chat.scrollHeight, behavior: "smooth"
    })
}

function getUserInfo(e, username) {
    // Avoid send primate message to current user
    if ((currentUser !== username) && (!arrayDiscussion.includes(e.dataset.channel_id))) {
        pm(username, e.dataset.channel_id)
        arrayDiscussion.push(e.dataset.channel_id)
        insertChat(e.dataset.channel_id);
    }
}

function insertChat(channel_id) {
    document.querySelector('#pills-tabContent').insertAdjacentHTML('beforeend', `<div class="tab-pane fade card-body" data-type="private_message" id="zone-chat-${channel_id}"></div>`)
}

function setChannel(e) {
    const notification = document.querySelector(`div#notification-` + e.dataset.channel_id);

    if (notification) {
        notification.remove()
    }

    localStorage.setItem('channel', e.dataset.channel_id);

    if (localStorage.getItem('channel') === 'tab-general') {
        localStorage.setItem('type', 'general');
    } else if (localStorage.getItem('channel') === 'pm') {
        localStorage.setItem('type', 'pm');
    } else {
        localStorage.setItem('type', 'room');
        let partyId = localStorage.getItem('party');
        let roomId = localStorage.getItem('room');
        insertChat("message_" + partyId + '_' + roomId);
    }
}

function insertTab(username, userId) {
    document.querySelector('#pills-tab').insertAdjacentHTML('beforeend', `<li class="nav-item" role="presentation">
            <button class="nav-link flex justify-center items-center px-1" id="${userId}" data-type="private_message" data-channel_id="${userId}" data-bs-toggle="pill" data-bs-target="#zone-chat-${userId}"
                    type="button" role="tab" aria-controls="pills-profile" aria-selected="false" onclick="setChannel(this)">` + username +
        '            </button>\n' +
        '        </li>')
}

function pm(username, userId) {
    localStorage.setItem('type', 'pm');
    insertTab(username, userId);
}

// Lorsqu'on envoie le formulaire, on transmet le message et on l'affiche sur la page
document.querySelector('#formulaire_chat').addEventListener('submit', function (e) {
    e.preventDefault();
    const message = document.querySelector('#message').value;
    let avatar = localStorage.getItem('avatar') ?? 'https://pkimgcdn.peekyou.com/dfd8bab38ab56d2b89dac1dad40b9e1e.jpeg';

    if (localStorage.getItem('type') === 'general') {
        // NEW PARAMS DONE
        sendMessage('tab-general', localStorage.getItem('party'), localStorage.getItem('room'), currentUser, avatar, message, userId, userId, 'tab-general');
    } else {
        // NEW PARAMS DONE
        sendMessage('private message', localStorage.getItem('party'), localStorage.getItem('room'), currentUser, avatar, message, userId, userId, 'pm');
    }

    // NEW PARAMS DONE
    insertMessage(localStorage.getItem('channel'), localStorage.getItem('partyId'), localStorage.getItem('room'), currentUser, avatar, message, userId, userId, 'message'); // Affiche le message aussi sur notre page
    document.querySelector('#message').value = ''; // Vide la zone de Chat et remet le focus dessus

    return false; // Permet de bloquer l'envoi "classique" du formulaire
});
